File: patch.py
import os
import requests
from flask import Flask, request, send_from_directory, jsonify
from urllib.parse import urljoin, urlparse
from datetime import datetime
import shutil

app = Flask(__name__)
app.config['X_FRAME_OPTIONS'] = 'SAMEORIGIN'

PORT = int(os.getenv('PORT', 5067))
TEMP_PATH = os.path.join(os.path.expanduser('~'), 'httptoolkit-patch')
APP_URL = f'http://localhost:{PORT}'

def request_url(method, url, redirect_count=0):
    global_proxy = os.getenv('GLOBAL_PROXY')
    proxies = {'https': global_proxy} if global_proxy else None

    try:
        response = requests.request(method, url, proxies=proxies, allow_redirects=False)
        if 300 <= response.status_code < 400 and 'Location' in response.headers:
            if redirect_count >= 5:
                raise Exception('Too many redirects')
            return request_url(method, response.headers['Location'], redirect_count + 1)
        return response
    except requests.RequestException as e:
        raise e

def has_internet():
    try:
        response = request_url('HEAD', 'https://app.httptoolkit.tech')
        return 200 <= response.status_code < 400
    except Exception:
        return False

@app.before_first_request
def setup():
    if not os.path.exists(TEMP_PATH):
        os.makedirs(TEMP_PATH)
    print(f'[Patcher] Selected temp path: {TEMP_PATH}')

@app.route('/<path:filename>', methods=['GET', 'HEAD'])
def serve_file(filename):
    print(f'[Patcher] Request to: {request.path}')
    file_path = os.path.join(TEMP_PATH, 'index.html' if request.path == '/' else request.path.lstrip('/'))
    if request.path in ['/view', '/intercept', '/settings', '/mock']:
        file_path += '.html'

    if request.path == '/ui-update-worker.js':
        return 'Not found', 404

    if not has_internet():
        print(f'[Patcher] No internet connection, trying to serve directly from temp path')
        if os.path.exists(file_path):
            print(f'[Patcher] Serving from temp path: {file_path}')
            return send_from_directory(TEMP_PATH, file_path)
        else:
            print(f'[Patcher] File not found in temp path: {file_path}')
            return 'No internet connection and file is not cached', 404

    try:
        if os.path.exists(file_path):
            try:
                remote_response = request_url('HEAD', urljoin('https://app.httptoolkit.tech', request.path))
                remote_date = datetime.strptime(remote_response.headers.get('Last-Modified', ''), '%a, %d %b %Y %H:%M:%S %Z')
                local_date = datetime.fromtimestamp(os.path.getmtime(file_path))
                if remote_date <= local_date:
                    print(f'[Patcher] File not changed, serving from temp path')
                    return send_from_directory(TEMP_PATH, file_path)
            except Exception as e:
                print(f'[Patcher] [ERR] Failed to fetch remote file date', e)
        
        print(f'[Patcher] File not found in temp path, downloading')
        remote_file = request_url('GET', urljoin('https://app.httptoolkit.tech', request.path))
        os.makedirs(os.path.dirname(file_path), exist_ok=True)
        data = remote_file.content

        if request.path == '/main.js':
            print(f'[Patcher] Patching main.js')
            response_headers = dict(remote_file.headers)
            response_headers['Cache-Control'] = 'no-store'

            data = data.decode()

            acc_store_name = re.search(r'class ([0-9A-Za-z_]+){constructor\(e\){this\.goToSettings=e', data)
            mod_name = re.search(r'([0-9A-Za-z_]+).(getLatestUserData|getLastUserData)', data)

            if acc_store_name:
                acc_store_name = acc_store_name.group(1)
            if mod_name:
                mod_name = mod_name.group(1)

            if not acc_store_name:
                print(f'[Patcher] [ERR] Account store name not found in main.js')
            elif not mod_name:
                print(f'[Patcher] [ERR] Module name not found in main.js')
            else:
                patched = data.replace(f'class {acc_store_name}{{', f'["getLatestUserData","getLastUserData"].forEach(p=>Object.defineProperty({mod_name},p{{value:()=>user}}));class {acc_store_name}{{')
                if patched == data:
                    print(f'[Patcher] [ERR] Patch failed')
                else:
                    patched = f'const user={json.dumps({{"email": "user@example.com", "subscription": {"status": "active", "expiry": "9999-12-31T00:00:00Z", "plan": "pro-annual"}}})};user.subscription.expiry=new Date(user.subscription.expiry);' + patched
                    data = patched
                    print(f'[Patcher] main.js patched')

        with open(file_path, 'wb') as f:
            f.write(data.encode() if isinstance(data, str) else data)
        print(f'[Patcher] File downloaded and saved: {file_path}')
        return send_from_directory(TEMP_PATH, file_path)
    except Exception as e:
        print(f'[Patcher] [ERR] Failed to fetch remote file: {file_path}', e)
        return 'Internal server error', 500

if __name__ == '__main__':
    app.run(port=PORT)

File: patch.js
const { HttpsProxyAgent } = require('https-proxy-agent');
const axios = require('axios').default;
const electron = require('electron');
const express = require('express');
const fs = require('fs');
const path = require('path');
const os = require('os');

function showPatchError(message) {
  console.error(message);
  electron.dialog.showErrorBox('Patch Error', `${message}\n\nPlease report this issue on GitHub (https://github.com/ThatNotEasy/HTTPToolkit-Patcher)`);
}

const axiosInstance = axios.create({
  baseURL: 'https://app.httptoolkit.tech',
  httpsAgent: globalProxy
    ? new HttpsProxyAgent(
        globalProxy.startsWith('http')
          ? globalProxy.replace(/^http:/, 'https:')
          : 'https://' + globalProxy
      )
    : undefined,
});

const hasInternet = () => axiosInstance.head('/').then(() => true).catch(() => false);

const port = process.env.PORT || 5067;
const tempPath = path.join(os.tmpdir(), 'httptoolkit-patch');

process.env.APP_URL = `http://localhost:${port}`;
console.log(`[Patcher] Selected temp path: ${tempPath}`);

const app = express();
app.disable('x-powered-by');

app.all('*', async (req, res) => {
  console.log(`[Patcher] Request to: ${req.url}`);
  
  let filePath = path.join(
    tempPath,
    new URL(req.url, process.env.APP_URL).pathname === '/' ? 'index.html' : new URL(req.url, process.env.APP_URL).pathname
  );
  
  if (['/view', '/intercept', '/settings', '/mock'].includes(new URL(req.url, process.env.APP_URL).pathname)) {
    filePath += '.html';
  }

  if (new URL(req.url, process.env.APP_URL).pathname === '/ui-update-worker.js') {
    return res.status(404).send('Not found');
  }

  if (!fs.existsSync(tempPath)) {
    fs.mkdirSync(tempPath);
  }

  if (!(await hasInternet())) {
    if (fs.existsSync(filePath)) {
      return res.sendFile(filePath);
    }
    return res.status(404).send('No internet connection and file is not cached');
  }

  try {
    if (fs.existsSync(filePath)) {
      const remoteDate = await axiosInstance.head(req.url).then((res) => new Date(res.headers['last-modified']));
      if (remoteDate < new Date(fs.statSync(filePath).mtime)) {
        return res.sendFile(filePath);
      }
    }

    const remoteFile = await axiosInstance.get(req.url, { responseType: 'arraybuffer' });
    for (const [key, value] of Object.entries(remoteFile.headers)) res.setHeader(key, value);

    const recursiveMkdir = (dir) => {
      if (!fs.existsSync(dir)) {
        recursiveMkdir(path.dirname(dir));
        fs.mkdirSync(dir);
      }
    };

    recursiveMkdir(path.dirname(filePath));
    let data = remoteFile.data;
    
    if (new URL(req.url, process.env.APP_URL).pathname === '/main.js') {
      data = data.toString();
      
      const accStoreName = data.match(/class ([0-9A-Za-z_$]+){constructor\(e\){this\.goToSettings=e/)?.[1];
      const modName = data.match(/([0-9A-Za-z_$]+).(getLatestUserData|getLastUserData)/)?.[1];
      
      if (!accStoreName) showPatchError(`[Patcher] [ERR] Account store name not found in main.js`);
      else if (!modName) showPatchError(`[Patcher] [ERR] Module name not found in main.js`);
      else {
        let patched = data.replace(
          `class ${accStoreName}{`,
          `["getLatestUserData","getLastUserData"].forEach(p=>Object.defineProperty(${modName},p,{value:()=>user}));class ${accStoreName}{`
        );
        
        if (patched !== data) {
          patched = `const user=${JSON.stringify({
            email, 
            subscription: {
              status: 'active',
              quantity: 1,
              expiry: new Date('9999-12-31').toISOString(),
              sku: 'pro-annual',
              plan: 'pro-annual',
              tierCode: 'pro',
              interval: 'annual',
              canManageSubscription: true,
              updateBillingDetailsUrl: 'https://github.com/ThatNotEasy/HTTPToolkit-Patcher',
            }
          })};user.subscription.expiry=new Date(user.subscription.expiry);` + patched;
          
          data = patched;
        } else showPatchError(`[Patcher] [ERR] Patch failed`);
      }
    }
    
    fs.writeFileSync(filePath, data);
    res.sendFile(filePath);
  } catch (e) {
    console.error(`[Patcher] [ERR] Failed to fetch remote file: ${filePath}`, e);
    res.status(500).send('Internal server error');
  }
});

app.listen(port, () => console.log(`[Patcher] Server listening on port ${port}`));

electron.app.on('ready', () => {
  electron.session.defaultSession.webRequest.onBeforeSendHeaders((details, callback) => {
    const blockedHosts = ['events.httptoolkit.tech'];
    if (blockedHosts.includes(new URL(details.url).hostname) || details.url.includes('sentry')) return callback({ cancel: true });
    details.requestHeaders.Origin = 'https://app.httptoolkit.tech';
    callback({ requestHeaders: details.requestHeaders });
  });
  electron.session.defaultSession.webRequest.onHeadersReceived((details, callback) => {
    details.responseHeaders['Access-Control-Allow-Origin'] = [`http://localhost:${port}`];
    callback({ responseHeaders: details.responseHeaders });
  });
});

electron.app.commandLine.appendSwitch('disable-http-cache');
